neuralmonkey/decoders/decoder.py

```diff
@@ -281,7 +303,49 @@ def basic_loop(previous_state, _):
 
             return self._dropout(input_embedding)
 
-        return basic_loop
+
+        def beamsearch_loop(beam_outputs, step):
+            """Beamsearch loop function.
+
+            Arguments:
+                beam_outputs: Beam-sized list of previous outputs
+                i: Maybe unused? argument
+            """
+            beam_size = len(beam_outputs)
+
+            # will store a list of batch x vocabulary tensors
+            softmaxes = []
+
+            for beam, output in enumerate(beam_outputs):
+                output_activation = self._logit_function(output)
+                beam_softmax = tf.nn.log_softmax(output_activation)
+                softmaxes.append(beam_softmax)
+
+            ## we sum the softmaxes to get probs of next word
+            ## given all histories
+            ## there some independence assumptions here, but who cares (?)
+
+            # beam x batch x vocabulary
+            packed_softmaxes = tf.pack(softmaxes)
+
+            # batch x vocabulary
+            summed_softmaxes = tf.reduce_sum(packed_softmaxes, 0)
```

Comment:
No je to nějaký zamotaný no.. Když jsem to psal tak jsem tak nějak cejtil že je to dobře, ale ten pocit je teď pryč. :-)